### PR TITLE
Correctly update layer control widget when visible layers change

### LINKED ISSUE #700: Calling `setVisibleLayers()` on a dynamic layer does not correctly update layer control widget
__How often can you reproduce it?__ 

- [x ] Always
- [ ] Sometimes
- [ ] Rarely
- [ ] Unable
- [ ] I didn’t try

__Description:__

Calling `setVisibleLayers()` on a dynamic layer does not correctly update the layer control widget.

__Steps to reproduce:__
1. Go to http://demo.cmv.io/viewer/
2. Expand all of the Louisvile Public Safety layers and sublayers
3. Untick all sublayers and folders under Louisvile Public Safety
4. Run this line of javascript: `app.layers[1].setVisibleLayers([]);` (that assumes that app.layers[1] is Louisvile Public Safety, the first layer didn't load - CORS blocked it I think)

__Expected results:__

All sublayers and folder under Louisvile Public Safety should be unticked and not visible.

__Actual results:__

All sublayers under Louisvile Public Safety are unticked and hidden, all folders are ticked.
It should also be possible to set some sublayers to be ticked in the layer control widget, but have their parent folder unticked (like how Schools under Public Safety is default unticked)

__Environment:__

| Software                | Version
| ------------------ | -------
| CMV Version | v2.0.0-beta.1
| Browser                 | Chrome 56.0.2924.87
| Operating system | Windows 10

### PR DESCRIPTION
**Pull request by duckblaster**
Correctly update layer control widget when calling `setVisibleLayers()` on a dynamic layer, fixes #700

### CODE REVIEW

File: viewer/js/gis/dijit/LayerControl/controls/Dynamic.js [hunk 1]
```diff
@@ -204,23 +208,46 @@ define([
...
+                        }
+                    });
+                } else if (info.subLayerIds !== null && array.indexOf(setLayers, info.id) !== -1) {
+                    setLayers.splice(array.indexOf(setLayers, info.id), 1);
                 }
             });
+            if (!setLayers.length) {
+                setLayers.push(-1);
+            }
+            return setLayers
```
Comment (green3g) [inline viewer/js/gis/dijit/LayerControl/controls/Dynamic.js]: Missing semicolon here

**Comment by tmcgee** [comment]
So based on the response from @roemhildtg on issue #700, there's nothing to do here, correct?

**Comment by duckblaster** [comment]
No, it's still broken, try 1,2,3 in the array instead

On Wed, 22 Mar 2017 at 6:35 AM, Tim McGee <notifications@github.com> wrote:

So based on the response from @roemhildtg <https://github.com/roemhildtg>
on issue #700 <https://github.com/cmv/cmv-app/issues/700>, there's nothing
to do here, correct?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/cmv/cmv-app/pull/702#issuecomment-288158026>, or mute
the thread

File: viewer/js/gis/dijit/LayerControl/controls/Dynamic.js [hunk 2]
```diff
@@ -204,23 +208,46 @@ define([
...
-                    visibleIds.push(info.id);
-                }
-                if (info.parentLayerId !== -1 && array.indexOf(visibleIds, info.parentLayerId) === -1) {
-                    visibleIds.push(info.parentLayerId);
+        _aspectSetVisibleLayers: function () {
+            var self = this;
+            this._visLayersHandler = aspect.around(this.layer, 'setVisibleLayers', lang.hitch(this, function (originalSetVisibleLayers) {
+                return function (visibleLayers) {
+                    var setLayers = lang.hitch(self, self._visibleLayersToSetLayers)(visibleLayers);
+                    originalSetVisibleLayers.apply(this, [setLayers]);
```
Comment (green3g) [inline viewer/js/gis/dijit/LayerControl/controls/Dynamic.js:221]: This line shouldn't pass the modified setLayers, it should pass the visibleLayers that the api expects to receive.

File: viewer/js/gis/dijit/LayerControl/controls/Dynamic.js [hunk 3]
```diff
@@ -204,23 +208,46 @@ define([
...
-            array.forEach(this.layer.layerInfos, function (info) {
-                if (array.indexOf(visLayers, info.id) !== -1) {
-                    visibleIds.push(info.id);
-                }
-                if (info.parentLayerId !== -1 && array.indexOf(visibleIds, info.parentLayerId) === -1) {
-                    visibleIds.push(info.parentLayerId);
+        _aspectSetVisibleLayers: function () {
+            var self = this;
+            this._visLayersHandler = aspect.around(this.layer, 'setVisibleLayers', lang.hitch(this, function (originalSetVisibleLayers) {
+                return function (visibleLayers) {
```
Comment (green3g) [inline viewer/js/gis/dijit/LayerControl/controls/Dynamic.js:219]: this should probably be `return lang.hitch(this, function(visibleLayers) {`

Comment (green3g) [reply viewer/js/gis/dijit/LayerControl/controls/Dynamic.js:219]: And eliminate the use of `self` (for consistency)

**Comment by tmcgee** [comment]
Now that PR #747 was __finally__ merged into the develop branch, it is time to revisit this PR and issue #700 . I'm trying to remember if it was addressed in #747. I am thinking it was (though done slightly different) but may be mistaken. @duckblaster @roemhildtg thoughts?

**Comment by duckblaster** [comment]
I think this PR can be scrapped, but we need a way to get and set the layer settings as shown to the user for the AppSettings widget, which is what this PR attempted to do.

**Comment by tmcgee** [comment]
@duckblaster thanks. I'll let you and @roemhildtg work through the next steps on improvements related to the AppSettings widget whether that is here in CMV core and/or an update to the widget.

**Pull request closed without merging**